python-frontend/src/test/java/org/sonar/python/tree/PythonTreeMakerTest.java

```diff
@@ -747,6 +749,54 @@ void nonlocalStatement() {
     assertThat(nonlocalStatement.children()).hasSize(3);
   }
 
+  @Test
+  void funcdef_statement_type_params() {
+    setRootRule(PythonGrammar.FUNCDEF);
+    var astNode = p.parse("def overly_generic[\n" +
+      "   SimpleTypeVar,\n" +
+      "   TypeVarWithBound: int,\n" +
+      "   TypeVarWithConstraints: (str, bytes),\n" +
+      "   *SimpleTypeVarTuple,\n" +
+      "   **SimpleParamSpec\n" +
+      "](\n" +
+      "   a: SimpleTypeVar,\n" +
+      "   b: TypeVarWithBound,\n" +
+      "   c: Callable[SimpleParamSpec, TypeVarWithConstraints],\n" +
+      "   *d: SimpleTypeVarTuple,\n" +
+      "): pass");
+    var functionDef = treeMaker.funcDefStatement(astNode);
+    assertThat(functionDef.name()).isNotNull();
+    var typeParams = functionDef.typeParams();
+    assertThat(typeParams).isNotNull();
+    assertThat(typeParams.getKind()).isEqualTo(Kind.TYPE_PARAMS);
+    assertThat(typeParams.leftBracket()).isNotNull();
+    assertThat(typeParams.rightBracket()).isNotNull();
+    assertThat(typeParams.children()).hasSize(11);
+    var typeParamsList = typeParams.typeParams();
+    assertThat(typeParamsList).isNotNull().hasSize(5);
+
+    var simpleTypeVar = typeParamsList.get(0);
+    assertThat(simpleTypeVar.name().name()).isEqualTo("SimpleTypeVar");
+    assertThat(simpleTypeVar.starToken()).isNull();
+    assertThat(simpleTypeVar.typeAnnotation()).isNull();
+
+    var typeWithBound = typeParamsList.get(1);
+    assertThat(typeWithBound.name().name()).isEqualTo("TypeVarWithBound");
+    assertThat(typeWithBound.starToken()).isNull();
+    assertThat(typeWithBound.typeAnnotation()).isNotNull();
+
+    var simpleTypeVarTuple = typeParamsList.get(3);
+    assertThat(simpleTypeVarTuple.name().name()).isEqualTo("SimpleTypeVarTuple");
+    assertThat(simpleTypeVarTuple.starToken()).isNotNull();
+    assertThat(simpleTypeVarTuple.typeAnnotation()).isNull();
+
+    var visitor = Mockito.spy(new BaseTreeVisitor());
+
+    functionDef.accept(visitor);
+    Mockito.verify(visitor, Mockito.times(1)).visitTypeParams(Mockito.any());
+    Mockito.verify(visitor, Mockito.times(5)).visitTypeParam(Mockito.any());
```

Comment:
Once using the `RuleTest#parse` method, maybe this is not needed anymore (same for the `ClassDef` version) and we can remove the `Mockito` import.

Reply:
Without it, it shows a lack of test coverage for these methods.

Reply:
When using `p.parse` only instead of `RuleTest#parse`, I can see why. But given that `RuleTest#parse` performs a visit of the tree, I'm confused why coverage would still be missing. Are you sure it's the case even with the updated method?